tools/ci/watchdog.sh

```diff
@@ -0,0 +1,122 @@
+#  Licensed to the Apache Software Foundation (ASF) under one
+#  or more contributor license agreements.  See the NOTICE file
+#  distributed with this work for additional information
+#  regarding copyright ownership.  The ASF licenses this file
+#  to you under the Apache License, Version 2.0 (the
+#  "License"); you may not use this file except in compliance
+#  with the License.  You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+#  Unless required by applicable law or agreed to in writing, software
+#  distributed under the License is distributed on an "AS IS" BASIS,
+#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#  See the License for the specific language governing permissions and
+# limitations under the License.
+################################################################################
+
+#
+# This file contains a watchdog tool to run monitor and potentially kill tasks
+# not producing any output for n seconds.
+#
+
+# Number of seconds w/o output before printing a stack trace and killing the watched process
+MAX_NO_OUTPUT=${MAX_NO_OUTPUT:-900}
+
+# Number of seconds to sleep before checking the output again
+SLEEP_TIME=${SLEEP_TIME:-20}
+
+CMD_OUT=${CMD_OUT:-"/tmp/watchdog.out"}
+CMD_PID=${CMD_PID:-"/tmp/watchdog.pid"}
+CMD_EXIT=${CMD_EXIT:-"/tmp/watchdog.exit"}
+
+
+# =============================================
+# Utility functions
+# ============================================= 
+
+mod_time () {
+	echo `stat -c "%Y" $CMD_OUT`
+}
+
+the_time() {
+	echo `date +%s`
+}
+
+# watchdog process
+
+watchdog () {
+	touch $CMD_OUT
+
+	while true; do
+		sleep $SLEEP_TIME
+
+		time_diff=$((`the_time` - `mod_time`))
+
+		if [ $time_diff -ge $MAX_NO_OUTPUT ]; then
+			echo "=============================================================================="
+			echo "Process produced no output for ${MAX_NO_OUTPUT} seconds."
+			echo "=============================================================================="
+
+			# run timeout callback
+			$WATCHDOG_CALLBACK_ON_TIMEOUT
+
+			echo "Killing process with pid=$(<$CMD_PID) and all descendants"
+			pkill -P $(<$CMD_PID) # kill descendants
+			kill $(<$CMD_PID) # kill process itself
+
+			exit 1
+		fi
+	done
+}
+
+assume_available () {
+	VAR=$1
+	if [ -z "$VAR" ] ; then
+		echo "ERROR: Environment variable '$VAR' is not set but expected by watchdog.sh"
+		exit 1
+	fi
+}
+
+# =============================================
+# main function
+# =============================================
+
+# entrypoint
+function run_with_watchdog() {
+	local cmd="$1"
+
+	# check preconditions
+	assume_available CMD_OUT # used for writing the process output (to check for activity)
+	assume_available CMD_PID # location of file to write process id to
+	assume_available CMD_EXIT # location of file to writ exit code to
+	assume_available WATCHDOG_CALLBACK_ON_TIMEOUT # bash function to call on timeout
+
+	watchdog &
```

Comment:
I'd prefer explicitly passing the required arguments.

Reply:
`watchdog` doesn't really have any required arguments. I see `SLEEP_TIME` and `MAX_NO_OUTPUT` more as configuration parameters.

Reply:
`CMD_PID` is **not** a variable containing the command process id. It is the filename used to store and retrieve the command process id from. In that sense, this should be more considered as an internal configuration, not an argument to be passed around.

There is probably no logical argument why the two approaches we are discussing are better or worse, because both approaches work.

But in my opinion, the readability is a bit better when globally defining the names of helper files, instead of globally defining them for the main method but passing it for helper methods as an argument. Something that is passed as an argument to a method is usually considered part of the actual work being done.

Reply:
Note that I also referred to the callback, which is very much not internal configuration.

Reply:
Thank you for your clarifications! I now understand your concerns. 
I agree that the use of `assume_available` for something internal is weird.

I will push an updated version of the script.